Data fixer performance - easy bits

Gegy opened the pull request:
Whenever the game is started up, DataFixerUpper builds and optimises functions that are able to upgrade data from every past data format to the most recent data format. Although the optimisation step comes at a high cost itself, it allows fixer functions, especially those that span many versions, to avoid unpacking/repacking data into intermediary structures when modifying deeply nested data many times. By building these at startup, this avoids latency during gameplay, but has a significant overhead during game launch. Although this process is extensively parallelised, it is one of the slowest parts of the game bootstrap process, and will only get slower with time as the data format continues to evolve.

This PR improves the performance of this process by addressing some easy-wins without changing any of the core logic for now.

Best reviewed commit-by-commit. Some commits have more in-depth explanations around specific changes.

## What is slow? Why is it slow?
While initially building the unoptimised functions is definitely a fairly time-consuming process, by far the cost of building datafixers comes from the optimisation step.
![image](https://user-images.githubusercontent.com/5172118/211612155-d22c60d3-bdbc-4970-8a52-3e0d610ccd4a.png)

In basic: the optimisation process works by applying various rules that restructure the produced datafixing functions based on the algebraic rules given by the point-free style of functions. Many of these rules are applied one-at-a-time (`many(once(...))` rule), where a single node in the point-free tree is rewritten before packing up the new modified tree. This particularly means that the same logic runs many, many times on the same data. The changes in this PR generally focus around reducing that additional traversal, as well as generally improving the performance of making that traversal.

~~The main change in this regard, and the highest-impact change in this PR, involves caching for individual branches of the tree when a given rule no longer applies. This avoids traversing already 'stable' parts of the tree.~~ This change is superseded by the change to no longer use the `once()` rule at all in #72.

Garbage collection is another aspect of this: the point-free optimisation step particularly generates a lot of short-lived objects that create high amounts of GC pressure.

|![image](https://user-images.githubusercontent.com/5172118/211606843-5a66e913-7d6f-4fd7-8f5f-b78666588dca.png)|![image](https://user-images.githubusercontent.com/5172118/211606624-c50e42ac-eacf-4bd9-96e9-41df321d3e2d.png)|
|-|-|

The majority these allocations come from the recursive point-free rewrite rules, which are flattened in this PR, as well as `Optional` usage and single-value types that were recreated for each use.

## Fixer application
While this PR mainly focuses on the bootstrap process, the last few commits also make a change for performance when applying fixers. A large amount of time spent in fixer application is spent calling `.toString()` on NBT data. This is due to the use of `Codec` and `DataResult`: when an error is encountered, the `DataResult` type can hold an error string to represent its failure. Most often, these errors include the stringified data that failed to be parsed. This wouldn't be a concern, because that data is needed to display the error, but these error strings are most often entirely discarded. This is, for example, frequently encountered through the `EitherCodec` which chooses one result and discards the error of the other if it fails.

## Results
From testing, these changes provide roughly a 7x improvement in datafixer bootstrap time, which is consistent both when running on a single thread as well as in parallel:

|Before|After|
|-|-|
|![image](https://user-images.githubusercontent.com/5172118/211610833-232eb2e2-7566-417f-8d0b-41b9a8f49083.png)|![image](https://user-images.githubusercontent.com/5172118/211609562-bad45afc-54ec-4785-8e49-9a38f360677b.png)|
|Average 35s in parallel, 136s on a single thread|Average 5s in parallel, 15s on a single thread|

For fixer application, the `DataResult` changes provide roughly a 1.8x improvement (before: 56s, after: 30s) to optimise a 1.6.4 world

Gegy commented:
Accidentally stumbled across another easy change: most of the fixers (roughly half) are optimised for types that are never used. Many types exist to represent an element contained within another type. For example, entities are contained within chunks: but an entity never needs to be updated alone. The chunk fixer function inlines the recursive functions that apply to entities, so the function that applies only to entities is never required.

This gives roughly a 2x speed-up on top of the existing changes.

bagucode reviewed (approved):
The code looks good to me and the speedup is amazing! But there are a lot of changes and I feel like it would have been easy to miss something during review so it would be good if we could somehow test this a lot before using this new, faster version in production.

Gegy commented:
@bagucode Thanks for looking through it! 🙂 We have run a basic consistency test on this PR and the follow-up by dumping inputs and outputs when upgrading an old world, and verifying against that with the new changes (discovered in the process one datafixer that generates random values 🥲)

Gegy commented:
Force-push: dropped change from optional -> nullable after feedback (the overhead is now negligible given changes in #72). Also dropped the `lastStableRule` cache change, as that is no longer used as of #72.

RainWarrior commented on src/main/java/com/mojang/datafixers/functions/PointFreeRule.java:
```diff
@@ -590,6 +595,32 @@ public int hashCode() {
+        }
+
+        @Override
+        public boolean equals(final Object obj) {
```
This is a record, so we probably do not need explicit equals/hashCode anymore :)

Gegy replied:
This one is unfortunately still required for the `Arrays.equals`/`Arrays.hashCode` :)

RainWarrior replied:
Ah, I see, default `.equals` does not do what we want for arrays; one more reason to prefer `List<T>` over `T[]`, but not that important here :)
